beacon-chain/sync/initial-sync/verification.go

```diff
@@ -0,0 +1,115 @@
+package initialsync
+
+import (
+	"context"
+
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/blockchain/kzg"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/das"
+	"github.com/prysmaticlabs/prysm/v4/beacon-chain/verification"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+)
+
+func newBlobVerifierFromInitializer(ini *verification.Initializer) verification.NewBlobVerifier {
+	return func(b blocks.ROBlob, reqs []verification.Requirement) verification.BlobVerifier {
+		return ini.NewBlobVerifier(b, reqs)
+	}
+}
+
+func newBlobBatchVerifier(newVerifier verification.NewBlobVerifier) *BlobBatchVerifier {
+	return &BlobBatchVerifier{
+		verified:    make(map[[32]byte]primitives.Slot),
+		verifyKzg:   kzg.Verify,
+		newVerifier: newVerifier,
+	}
+}
+
+type kzgVerifier func(b ...blocks.ROBlob) error
+
+// BlobBatchVerifier solves problems that come from verifying batches of blobs from RPC.
+// First: we only update forkchoice after the entire batch has completed, so the n+1 elements in the batch
+// won't be in forkchoice yet.
+// Second: it is more efficient to batch some verifications, like kzg commitment verification. Batch adds an additional
+// method to BlobVerifier to verify the kzg commitments of all blob sidecars for a block together, then using the cached
+// result of the batch verification when verifying the individual blobs.
+type BlobBatchVerifier struct {
+	verifyKzg   kzgVerifier
+	newVerifier verification.NewBlobVerifier
+	verified    map[[32]byte]primitives.Slot
+}
+
+var _ das.BlobBatchVerifier = &BlobBatchVerifier{}
+
+func (kb *BlobBatchVerifier) VerifiedROBlobs(ctx context.Context, scs []blocks.ROBlob) ([]blocks.VerifiedROBlob, error) {
+	if err := kb.verifyKzg(scs...); err != nil {
+		return nil, err
+	}
+	vs := make([]blocks.VerifiedROBlob, len(scs))
+	for i := range scs {
+		vb, err := kb.verifyOneBlob(ctx, scs[i])
+		if err != nil {
+			return nil, err
+		}
+		vs[i] = vb
+	}
+	return vs, nil
+}
+
+func (kb *BlobBatchVerifier) verifyOneBlob(ctx context.Context, sc blocks.ROBlob) (blocks.VerifiedROBlob, error) {
+	vb := blocks.VerifiedROBlob{}
+	bv := kb.newVerifier(sc, verification.InitsyncSidecarRequirements)
+	// We can satisfy this immediately because VerifiedROBlobs always verifies commitments for all blobs in the batch
+	// before calling verifyOneBlob.
+	bv.SatisfyRequirement(verification.RequireSidecarKzgProofVerified)
+
+	if err := bv.BlobIndexInBounds(); err != nil {
+		return vb, err
+	}
+	if err := bv.NotFromFutureSlot(); err != nil {
+		return vb, err
+	}
+	if err := bv.SlotAboveFinalized(); err != nil {
+		return vb, err
+	}
+	if err := bv.ValidProposerSignature(ctx); err != nil {
+		return vb, err
+	}
```

Comment:
I'm a bit concerned that the order of these checks isn't optimal.

Reply:
should be resolved here: https://github.com/prysmaticlabs/prysm/pull/13351/commits/9ade4cf800439d243945fb7ab79fbb06264ff104